services/src/datasets/external/netcdfcf/mod.rs

```diff
@@ -840,103 +855,270 @@ impl DataProvider for NetCdfCfDataProvider {
+        NetCdfCfDataProvider::build_netcdf_tree(&provider_path, Some(&overview_path), &fp, false)
+            .map_err(|_| Error::InvalidLayerCollectionId)
+    })
+    .await??;
+
+    let groups_list = find_group(tree.groups.clone(), groups)?.map_or(tree.groups, |g| g.groups);
+
+    if groups_list.is_empty() {
+        tree.entities
+            .into_iter()
+            .skip(options.offset as usize)
+            .take(options.limit as usize)
+            .map(|entity| {
+                Ok(CollectionItem::Layer(LayerListing {
+                    id: ProviderLayerId {
+                        provider_id: NETCDF_CF_PROVIDER_ID,
+                        layer_id: NetCdfLayerCollectionId::Entity {
+                            path: file_path.clone(),
+                            groups: groups.to_owned(),
+                            entity: entity.id,
+                        }
+                        .try_into()?,
+                    },
+                    name: entity.name,
+                    description: "".to_string(),
+                    properties: None,
+                }))
+            })
+            .collect()
+    } else {
+        let out_groups = groups.to_owned();
+
+        groups_list
+            .into_iter()
+            .skip(options.offset as usize)
+            .take(options.limit as usize)
+            .map(|group| {
+                let mut out_groups = out_groups.clone();
+                out_groups.push(group.name.clone());
+                Ok(CollectionItem::Collection(LayerCollectionListing {
+                    id: ProviderLayerCollectionId {
+                        provider_id: NETCDF_CF_PROVIDER_ID,
+                        collection_id: NetCdfLayerCollectionId::Group {
+                            path: file_path.clone(),
+                            groups: out_groups,
+                        }
+                        .try_into()?,
+                    },
+                    name: group.title.clone(),
+                    description: group.description,
+                    entry_label: None,
+                    properties: None,
+                }))
+            })
+            .collect()
+    }
+}
+
 #[async_trait]
 impl LayerCollectionProvider for NetCdfCfDataProvider {
     async fn collection_items(
         &self,
         collection: &LayerCollectionId,
         options: Validated<LayerCollectionListOptions>,
     ) -> crate::error::Result<Vec<CollectionItem>> {
-        ensure!(
-            *collection == self.root_collection_id().await?,
-            crate::error::UnknownLayerCollectionId {
-                id: collection.clone()
-            }
-        );
+        let id: NetCdfLayerCollectionId = serde_json::from_str(&collection.0)?;
 
-        let mut dir = tokio::fs::read_dir(&self.path).await?;
-
-        let mut datasets = vec![];
-        while let Some(entry) = dir.next_entry().await? {
-            if !entry.path().is_file() {
-                continue;
+        match id {
```

Comment:
Do you check somewhere that you cannot escape the NetCDF path?
E.g., that you cannot provide `../../hidden.nc` or `/media/secret/hidden.nc`?

I did this at some points by checking the the absolute file path I will load has the absolute provider path as its prefix or something like this.

Reply:
sounds like a good idea. Can you point me to the code that did this, because I couldn't find it.

Reply:
if you mean this: https://github.com/geo-engine/geoengine/blob/c2406c742467d6f002ab0ab7bc2e2715868ccf9f/services/src/datasets/external/netcdfcf/mod.rs#L399 then either I don't understand it or it doesn't work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fcbe04de9cbaacd7bcc36f4fe7051bec

Reply:
Good point, we could use `join`, then `canonicalize`, and then check for `starts_with`?

Reply:
9d0604e03b5a4ad399543e26b02c0f5a2490073b